### PR TITLE
Free modules, submodules and quotient modules

### PR DESCRIPTION
**Pull request by wbhart**
Claus suggested I make a PR to be merged now, so that this doesn't bitrot over the next month.

* Currently does not compute retraction/section when constructing maps for quotient/submodules
* QuotientModule should probably return the canonical projection for the forward direction, instead of the lift as currently
* Test code is not randomised (but passes)
* Quotient modules only able to be constructed from modules that are actual submodules

### CODE REVIEW

**Comment by thofma** [comment]
Thanks. Docs are not building though.

**Comment by wbhart** [comment]
@thofma Should be ok now.

File: src/AbstractAlgebra.jl [hunk 1]
```diff
@@ -416,6 +418,56 @@ function NumberField(a::AbstractAlgebra.Generic.Poly{Rational{BigInt}}, s::Char,
    NumberField(a, string(s), t; cached=cached)
 end
 
+function FreeModule(R::NCRing, rank::Int; cached::Bool = true)
```
[the remark this answers is not in the document]
Comment (wbhart) [reply src/AbstractAlgebra.jl]: Fixed.

File: src/AbstractAlgebra.jl [hunk 2]
```diff
@@ -416,6 +418,56 @@ function NumberField(a::AbstractAlgebra.Generic.Poly{Rational{BigInt}}, s::Char,
...
+   Generic.FreeModule(R, rank; cached=cached)
+end
+
+function VectorSpace(R::Field, dim::Int; cached::Bool = true)
```
[the remark this answers is not in the document]
Comment (wbhart) [reply src/AbstractAlgebra.jl]: Fixed.

**Comment by wbhart** [comment]
@heiderich I've made the FreeModule constructor (which both of these use) now reject negative Int ranks. (It's not appropriate to use UInt here, since UInts are really for bit twiddling and shouldn't be required to do mathematics.)

Currently this implementation of modules is only an early sketch, with almost all nontrivial functionality missing. So I wouldn't expect too much from it just now (including correctness).

**Pull request merged**